File: Baekjoon/Problem_12000~12999/Problem_12015/main.cpp
#include <cstdio>
#include <vector>

using namespace std;

int binary_search(const vector<int> &sequence, const int s, const int e, const int target) {
    // base
    if (s >= e) return s;

    // default
    const int m = (s + e) / 2;
    if (sequence[m] >= target) return binary_search(sequence, s, m, target);
    else return binary_search(sequence, m+1, e, target);
}

int solve(const vector<int> &sequence) {
    vector<int> lis;

    for (auto e: sequence) {
        // 마지막 원소보다 큰 경우
        if (lis.empty() || e > lis.back()) {
            lis.push_back(e);
        } else {
            const int index = binary_search(lis, 0, lis.size(), e);
            lis[index] = e;
        }
    }

    return lis.size();
}

int main() {
    int N; scanf("%d", &N);
    vector<int> A(N); for (int i = 0; i < N; i++) scanf("%d", &A[i]);

    printf("%d\n", solve(A));

    return 0;
}

File: Baekjoon/Problem_12000~12999/Problem_12015/main.py
import sys

input = lambda: sys.stdin.readline().rstrip()

def binary_search(sequence: list[int], s: int, e: int, target: int) -> int:
    # base
    if s >= e: return s
    
    # default
    m = (s + e) // 2
    if sequence[m] >= target:
        return binary_search(sequence, s, m, target)
    else:
        return binary_search(sequence, m+1, e, target)

def solve(sequence: list[int]):
    lis = []
    
    for e in sequence:
        # 처음 삽인인 경우 or 더 큰 원소가 들어오는 경우
        if not lis or e > lis[-1]:
            lis.append(e)
        # 일반적인 경우 교체한다.
        else:
            index = binary_search(lis, 0, len(lis)-1, e)
            lis[index] = e
    
    return len(lis)

if __name__ == "__main__":
    N = int(input())
    A = list(map(int, input().split()))

    print(solve(A))